[MRG] CSS: improve rendering of download buttons

GaelVaroquaux opened the pull request:
This PR:

* Fixes the overflow visible in the download buttons (as discussed in https://github.com/scikit-learn/scikit-learn/pull/7354#issuecomment-246689940)

* Centers the buttons correctly

* Make the sphinx-gallery signature slightly less visible (let's not annoy downstream users)

Before, in sphinx-gallery's documentation:

![image](https://cloud.githubusercontent.com/assets/208217/18550236/0cc44b3c-7b53-11e6-8870-eddac36c7ab9.png)

After:

![image](https://cloud.githubusercontent.com/assets/208217/18550192/df186876-7b52-11e6-976c-e235246140f6.png)

Before in scipy-lectures
![image](https://cloud.githubusercontent.com/assets/208217/18550320/83f024ec-7b53-11e6-860c-dda162295215.png)

After:
![image](https://cloud.githubusercontent.com/assets/208217/18550347/a52982e8-7b53-11e6-8ad3-bf2581ec6e20.png)

Titan-C commented:
I would say some buttons become really small. 
![gal](https://cloud.githubusercontent.com/assets/713451/18552923/de74d060-7b5e-11e6-914a-129e80a5c1c8.png)

I would put a min-width instead of completely deleting the max-width. Otherwise the button look too small and we might loose visual uniformity.

Titan-C commented on sphinx_gallery/_static/gallery.css:
```diff
@@ -143,8 +150,13 @@ ul.sphx-glr-horizontal img {
+  margin-left: auto;
+  display: table;
+}
+
```
You can put 
margin-left: auto;
display: table;

in the next class and you get the same effect without having this separate class

Titan-C commented on sphinx_gallery/_static/gallery.css:
```diff
@@ -115,12 +120,14 @@ div.sphx-glr-download a {
+
 div.sphx-glr-download code.download {
   display: inline-block;
 }
```
To solve the conflict for scikit-learn with the boostrap.min.css we need to add an extra `white-space: normal;` in this class. Or shall that be kept only in scikit-learn css files?

GaelVaroquaux commented:
> To solve the conflict for scikit-learn with the boostrap.min.css we
> need to add an extra white-space: normal; in this class. Or shall that
> be kept only in scikit-learn css files?

This sounds something reasonnable. And other people might be using
bootstrap. I'll add it to the PR.

GaelVaroquaux commented:
I've addressed all the comments, in particular the one on the width of the buttons. It now renders as such:
![image](https://cloud.githubusercontent.com/assets/208217/18557204/523ff4d4-7b6e-11e6-94ce-e16b4ebb06fd.png)

Titan-C commented:
I tested this PR on top of my new PR #147 that allows to change themes on the fly. It looks good and uniform on the classic theme and also the read the docs theme.

But one last nitpick. In the nature theme which is at the base for nilearn and scikit-learn you get this quite often
![image](https://cloud.githubusercontent.com/assets/713451/18619352/6b4a8faa-7dfa-11e6-9848-89128bfab708.png)

The jupyter notebook is usually 3 lines height and the Python source is 2. The boxes are unequal. Should we care?  adding `min-height: 3em;` might be enough `div.sphx-glr-download a` to give some extra uniformity.

GaelVaroquaux commented:
Hum, my fix doesn't fix what you had noticed above, as it fixes only for the download in the gallery. Could you please give me the steps to reproduce? I haven't managed.

Titan-C commented:
```bash
git checkout -b css_download_multitheme master
git pull https://github.com/Titan-C/sphinx-gallery.git sphinx-themes
git pull https://github.com/GaelVaroquaux/sphinx-gallery.git css_download
cd doc
make html theme=nature
```
This will, in a new branch, load my changes, then yours. On that branch you can build the docs under any theme, nature is the case of my snapshot. Test over the actual examples, `just_code.py` in the snapshot

lesteve commented:
The PR with themes has been merged into master so you just need to rebase on master to be able to test with the nature theme.

GaelVaroquaux commented:
> The PR with themes has been merged into master so you just need to rebase on master to be able to test with the nature theme.

Yes, I got it, and I can reproduce. I just haven't had time. Maybe this
morning.

GaelVaroquaux commented:
I've pushed a change that does vertical alignment:

![image](https://cloud.githubusercontent.com/assets/208217/18660587/95c20e30-7f11-11e6-82f1-bd0c3ebb0e6b.png)

GaelVaroquaux commented:
I've just pushed one extra set of changes that means that if the screen is wide enough, the buttons will be wider and they won't be 3 lines-long. On my screen:

![image](https://cloud.githubusercontent.com/assets/208217/18661210/d8d76e7e-7f14-11e6-9e26-f416b4d4ccf2.png)

Titan-C commented:
Great! 
But again a new Nitpick. I see in the read the docs theme
![buttons](https://cloud.githubusercontent.com/assets/713451/18668322/1976381e-7f35-11e6-9b38-506dd2259ab3.png)

When the file name is just too long, and because this boxes are `code` html tags. The read the docs puts `overflow-x: auto;` to them. Should we add to default `div.sphx-glr-download code.download` an extra `overflow-x: hidden;`? or put that into our theme overlay?

GaelVaroquaux commented:
Reproduced and fixed. I also removed the inner bevel of the buttons created by the RTD theme:
![image](https://cloud.githubusercontent.com/assets/208217/18669936/a7d00df8-7f3d-11e6-9632-a1e117d3e33c.png)

Titan-C commented:
I see you are cutting the notebook extension into `ipyn` and `b`

GaelVaroquaux commented:
> I see you are cutting the notebook extension into ipyn and b

Yes, because the word is too long, the browser needs to cut it.

Titan-C commented:
I was thinking it we need to have the file name in the button at all? The original reason the file names where there is to declare the download link. But now we have buttons. So maybe to keep visual uniformity  just have download buttons with out writing the filename on it. That will make it easier too, to setup the CSS styles.
It is then the user's browser which will uncover the filename upon clicking.

GaelVaroquaux commented:
The reason to have the filename is that people download the file, it then
goes in a huge directory with many many files downloaded (eg ~/Downloads
under Ubuntu) and people don't know which one it is. Of course, they can
find this out by looking at the "download" list of their browser, but it
makes a case for struggles and errors.

lesteve commented:
I know very little about CSS but it seems very hard to have the download buttons render well in many different themes (and I am guessing it depends a lot of the number of characters in the filename ...). I feel it is turning a little bit into bikeshedding.

I also feel we should prioritise the one on readthedocs (since this is our doc) and maybe the default sphinx theme. For other themes I am afraid people will have to tweak their CSS if they want to achieve perfection.

Having said that I am fine merging this as is and revisit later if it turns out to be too visually annoying.

GaelVaroquaux commented:
OK, you're pushing my web skills to their limits!

Here is an improvement: I am using the fact that the last element of the download button is the file name to make it smaller: screenshots, first with the nature theme, and then with the rtd theme

![image](https://cloud.githubusercontent.com/assets/208217/18705063/f3d28e60-7fec-11e6-9d06-59302c6c59bb.png)

![image](https://cloud.githubusercontent.com/assets/208217/18705080/121d83a2-7fed-11e6-8ebb-0c4ceefa2926.png)

![image](https://cloud.githubusercontent.com/assets/208217/18705103/2edbc7ec-7fed-11e6-8b72-efca4d59c50f.png)

![image](https://cloud.githubusercontent.com/assets/208217/18705144/50cce372-7fed-11e6-9b78-f62c51449723.png)

GaelVaroquaux commented:
I only partly agree with @lesteve. In a sense, yes it's going to be impossible to get everything perfect for every theme. However, people tend not to be very good at CSS, so the more work we do, the more they will love us. In addition, we need to have flashy selling points, and if using sphinx-gallery makes a gallery look awesome with no work, people will feel more driven to adopt us.

To illustrate that the current CSS works quite well with yet another theme, here is how it looks in the scipy-lectures:

![image](https://cloud.githubusercontent.com/assets/208217/18705413/7aa4e978-7fee-11e6-866c-ca6d13e1f8d7.png)

![image](https://cloud.githubusercontent.com/assets/208217/18705446/8d7f9a2a-7fee-11e6-890b-976f4a114f64.png)

Titan-C commented:
> Of course, they can find this out by looking at the "download" list of their browser, but it makes a case for struggles and errors.

Usually the browser handles that very well listing your last download.

The smaller fonts work well enough for me. 

+1

GaelVaroquaux commented:
>     Of course, they can find this out by looking at the "download" list of their browser, but it makes a case for struggles and errors.

> Usually the browser handles that very well listing your last download.

My comment is not a theoretical one. It's an experimental observation
based on teaching many classes to non technical people (eg master
students in psychology) who don't know well how to use a computer.

lesteve commented:
@Titan-C gave its +1, so let's merge this one!

A for effort on this one @GaelVaroquaux !
